### PR TITLE
AUTO-677 Create zero-downtime autoscaling pattern

### PR DESCRIPTION
**Pull request by GypsyBojangles**
* Create zero-downtime autoscaling class
* Create network, elb, database and asg config objects to reduce excessive parameter passing
* Update Cerberus schema to reflect new nested config objects and use yaml anchors to reduce duplication in Cerberus yaml.
* Cerberus schema is no longer changeable, has been moved to schema folder (anticipating new schemas for some upcoming new stack patterns)
* Updates to unit and system tests
* Updates to yaml defaults

Passing builds:
https://gaautobots.atlassian.net/builds/browse/AM-AUT51 - unit tests
https://gaautobots.atlassian.net/builds/browse/AM-AST43 - stack tester

### CODE REVIEW

File: amazonia/classes/util.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import yaml
+
```
Comment (fuddstar) [inline amazonia/classes/util.py]: import logging 

# Logging
    logging.basicConfig(
        level=logging.INFO
    )

File: amazonia/classes/yaml.py [hunk 2]
```diff
@@ -3,15 +3,51 @@
 Ingest User YAML and defaults YAML
 Overwrite defaults YAML with User YAML
 """
+import os
 import re
+
```
Comment (fuddstar) [inline amazonia/classes/yaml.py]: New line required?

File: amazonia/amz.py [hunk 3]
```diff
@@ -34,8 +24,8 @@ def create_stack(united_data):
     return stack
 
 
-def generate_template(yaml_data, default_data, schema_data):
-    yaml_return = Yaml(yaml_data, default_data, schema_data)
+def generate_template(yaml_data, default_data):
+    yaml_return = Yaml(yaml_data, default_data)
```
Comment (fuddstar) [inline amazonia/amz.py]: Docstrings

File: amazonia/classes/yaml.py [hunk 4]
```diff
@@ -94,31 +119,55 @@ def set_unit_values_for_type(self, unit_type):
...
-                    self.detect_unencrypted_access_keys(self.united_data[unit_type][unit]['userdata'])
-                # Validate that minsize is less than maxsize
-                if unit_value == 'minsize':
-                    minsize = self.united_data[unit_type][unit][unit_value]
-                    maxsize = self.user_stack_data[unit_type][unit].get('maxsize', self.default_data['maxsize'])
-                if minsize > maxsize:
-                    raise cerberus.ValidationError('Autoscaling unit minsize ({0}) cannot be '\
-                                                   'larger than maxsize ({1})'.format(minsize, maxsize))
 
+    def set_nested_object_values(self, nested_object_user_data, nested_object_default, nested_object_type):
```
Comment (fuddstar) [inline amazonia/classes/yaml.py]: Docstrings

File: amazonia/classes/yaml.py [hunk 5]
```diff
@@ -3,15 +3,51 @@
...
+
 import cerberus
 
+from amazonia.classes.util import read_yaml
+
 
 class Yaml(object):
     """
     Setting these as class variables rather than instance variables so that they can be resolved and referred to
     statically
```
Comment (fuddstar) [inline amazonia/classes/yaml.py]: Recommend docstrings for each keylists as if they were params

File: amazonia/classes/yaml.py [hunk 6]
```diff
@@ -94,31 +119,55 @@ def set_unit_values_for_type(self, unit_type):
...
+                minsize = unified_object[object_key]
+                maxsize = nested_object_user_data.get('maxsize', nested_object_default['maxsize'])
+            if minsize > maxsize:
+                raise cerberus.ValidationError('Autoscaling unit minsize ({0}) cannot be ' \
+                                               'larger than maxsize ({1})'.format(minsize, maxsize))
+        return unified_object
 
     @staticmethod
     def validate_yaml(data, schema):
-
```
Comment (fuddstar) [inline amazonia/classes/yaml.py]: Docstrings

File: amazonia/classes/asg_config.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        self.userdata = userdata
+        self.iam_instance_profile_arn = iam_instance_profile_arn
+        self.hdd_size = hdd_size
+
+    def define_undefined_values(self, other_asg_conifg):
+        """
+        Method to override undefined values with values from another ASGConfig class
+        :param other_asg_conifg:
+        """
+        self.sns_topic_arn = other_asg_conifg.sns_topic_arn if self.sns_topic_arn is None else self.sns_topic_arn
```
Comment (aynsof) [inline amazonia/classes/asg_config.py]: Typo - should be 'other_asg_config'

File: amazonia/classes/zd_autoscaling_unit.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        :param network_config: network related paramters including subnet, nat, jump, etc
+        :param elb_config: shared elb configuration
+        :param common_asg_config: default asg configuration
+        :param blue_asg_config: configuration specific to the blue asg
+        :param green_asg_config: configuration specific to the green asg
+        """
+        self.template = template
+        self.dependencies = dependencies if dependencies else []
+        self.elb_config = elb_config
+        self.prod_elb = Elb(
```
Comment (fuddstar) [inline amazonia/classes/zd_autoscaling_unit.py]: Add 'Create ELB' Comment

Comment (fuddstar) [inline amazonia/classes/zd_autoscaling_unit.py]: Add 'Create ELB' Comment

File: amazonia/classes/single_instance.py [hunk 9]
```diff
@@ -1,35 +1,26 @@
 #!/usr/bin/python3
+#  pylint: disable=line-too-long
```
Comment (aynsof) [inline amazonia/classes/single_instance.py]: Necessary?

File: amazonia/classes/zd_autoscaling_unit.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            elb_config=elb_config
+        )
+
+        blue_load_balancers = []
+        green_load_balancers = []
+
+        blue_load_balancers.append(self.prod_elb.trop_elb)
+        green_load_balancers.append(self.test_elb.trop_elb)
+
+        blue_asg_config.define_undefined_values(common_asg_config)
```
Comment (fuddstar) [inline amazonia/classes/zd_autoscaling_unit.py]: Add 'Create ASG' Comment

File: amazonia/classes/zd_autoscaling_unit.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        )
+        self.green_asg = Asg(
+            title='green' + unit_title,
+            template=self.template,
+            load_balancers=green_load_balancers,
+            network_config=network_config,
+            asg_config=green_asg_config
+        )
+
+        [self.prod_elb.add_flow(receiver=self.blue_asg, port=instanceport)
```
Comment (fuddstar) [inline amazonia/classes/zd_autoscaling_unit.py]: Add 'Creating Flow' Comment

File: amazonia/classes/autoscaling_unit.py [hunk 12]
```diff
@@ -1,106 +1,55 @@
 #!/usr/bin/python3
-#  pylint: disable=too-many-arguments, line-too-long
```
Comment (fuddstar) [inline amazonia/classes/autoscaling_unit.py]: Remove new line?

File: amazonia/classes/elb.py [hunk 13]
```diff
@@ -1,73 +1,68 @@
 #!/usr/bin/python3
 
-from amazonia.classes.security_enabled_object import SecurityEnabledObject
-from troposphere import Tags, Ref, Output, Join, GetAtt, route53
 import troposphere.elasticloadbalancing as elb
+from troposphere import Tags, Ref, Output, Join, GetAtt, route53
+
```
Comment (fuddstar) [inline amazonia/classes/elb.py]: Remove new line?

File: amazonia/classes/stack.py [hunk 14]
```diff
@@ -156,39 +175,74 @@ def __init__(self, stack_title, code_deploy_service_role, keypair, availability_
...
+        for unit in self.zd_autoscaling_units:  # type: dict
+            orig_unit_title = unit['unit_title']
+            if orig_unit_title in self.units:
+                raise DuplicateUnitNameError("Error: zd_autoscaling unit name '{0}' has already been specified, "
+                                             'it must be unique.'.format(orig_unit_title))
+            # Update unit title with stackname prefix
+            unit['unit_title'] = self.title + orig_unit_title
+            elb_config = ElbConfig(**unit['elb_config'])
+            common_asg_config = AsgConfig(keypair=self.keypair,
+                                          cd_service_role_arn=self.code_deploy_service_role,
```
Comment (fuddstar) [inline amazonia/classes/stack.py]: Move to network config

File: amazonia/classes/stack.py [hunk 15]
```diff
@@ -156,39 +175,74 @@ def __init__(self, stack_title, code_deploy_service_role, keypair, availability_
...
+                                          **unit['common_asg_config']
+                                          )
+            blue_asg_config = AsgConfig(keypair=self.keypair,
+                                        cd_service_role_arn=self.code_deploy_service_role,
```
Comment (fuddstar) [inline amazonia/classes/stack.py]: Move to network config

File: amazonia/classes/stack.py [hunk 16]
```diff
@@ -156,39 +175,74 @@ def __init__(self, stack_title, code_deploy_service_role, keypair, availability_
...
+                                        **unit['blue_asg_config']
+                                        )
+            green_asg_config = AsgConfig(keypair=self.keypair,
+                                         cd_service_role_arn=self.code_deploy_service_role,
```
Comment (fuddstar) [inline amazonia/classes/stack.py]: Move to network config

File: amazonia/classes/zd_autoscaling_unit.py [hunk 17]
```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/python3
+#  pylint: disable=too-many-arguments, line-too-long
+
+from amazonia.classes.asg import Asg
+from amazonia.classes.elb import Elb
+
+
+class ZdAutoscalingUnit(object):
+    def __init__(self, unit_title, template, dependencies, network_config, elb_config, common_asg_config,
```
Comment (fuddstar) [inline amazonia/classes/zd_autoscaling_unit.py]: Remove common

File: amazonia/classes/asg_config.py [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+        self.keypair = keypair
+        self.minsize = minsize
+        self.maxsize = maxsize
+        self.image_id = image_id
+        self.instance_type = instance_type
+        self.userdata = userdata
+        self.iam_instance_profile_arn = iam_instance_profile_arn
+        self.hdd_size = hdd_size
+
+    def define_undefined_values(self, other_asg_config):
```
Comment (fuddstar) [inline amazonia/classes/asg_config.py]: Remove undefined

File: amazonia/defaults.yaml [hunk 19]
```diff
@@ -27,59 +27,67 @@ home_cidrs:
...
+  image_id: 'ami-dc361ebf'
+  instance_type: 't2.micro'
+  iam_instance_profile_arn:
+  minsize: '1'
+  maxsize: '1'
+  health_check_grace_period: '300'
+  health_check_type: 'ELB'
+  hdd_size:
+  sns_topic_arn:
+  sns_notification_types:
```
Comment (aynsof) [inline amazonia/defaults.yaml]: Add sns_notification_types defaults:
      - 'autoscaling:EC2_INSTANCE_LAUNCH'
      - 'autoscaling:EC2_INSTANCE_LAUNCH_ERROR'
      - 'autoscaling:EC2_INSTANCE_TERMINATE'
      - 'autoscaling:EC2_INSTANCE_TERMINATE_ERROR'

**Comment by aynsof** [comment]
👍 

[Approved with PullApprove](https://pullapprove.com/GeoscienceAustralia/amazonia/pull-request/72/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=GeoscienceAustralia/amazonia)

**Pull request merged**